### PR TITLE
sgx-shim: Add unit tests for Heap

### PR DESCRIPTION
**Pull request by ueno**
This adds a couple of unit tests that exercise the methods (mmap, munmap, and brk) of Heap to ensure that the object state is properly updated.  This also fixes an issue that munmap returns an error if the mapped memory ends with the last page in the heap.

### CODE REVIEW

File: enarx-keep-sgx-shim/src/heap.rs [hunk 1]
```diff
@@ -200,3 +204,156 @@ impl Heap {
...
+            (Page::size() + Page::size() / 2, 2),
+            (pages * Page::size(), pages),
+        ];
+
+        for (s, allocated) in steps.iter() {
+            let addr = heap.mmap(
+                0,
+                *s,
+                libc::PROT_READ as _,
+                (libc::MAP_PRIVATE | libc::MAP_ANONYMOUS) as _,
```
Comment (npmccallum) [inline enarx-keep-sgx-shim/src/heap.rs:257]: There is a lot of wrapping in this file due to these two constants. Could you perhaps make some global `const`s?

File: enarx-keep-sgx-shim/src/heap.rs [hunk 2]
```diff
@@ -200,3 +204,156 @@ impl Heap {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    const ERRNO_BASE: usize = !0xfff;
+
+    unsafe fn new_checked(bytes: &mut [u8]) -> Heap {
+        let aligned = slice(bytes.as_mut_ptr(), bytes.len())
+            .align_to_mut::<Page>()
+            .1;
```
Comment (npmccallum) [inline enarx-keep-sgx-shim/src/heap.rs:216]:
```suggestion
        let aligned = bytes.align_to_mut::<Page>().1;
```

File: enarx-keep-sgx-shim/src/heap.rs [hunk 3]
```diff
@@ -200,3 +204,156 @@ impl Heap {
...
+            count: (aligned.len() * Page::size()) as _,
+        };
+        let heap = Heap::new(span);
+
+        // brk points to the first page
+        assert_eq!(heap.metadata.brk.start, heap.pages.as_ptr() as _);
+        assert_eq!(heap.metadata.brk.end, heap.pages.as_ptr() as _);
+
+        // exclude 2 pages for metadata and allocation map
+        assert_eq!(heap.pages.len(), pages - 2);
```
Comment (npmccallum) [inline enarx-keep-sgx-shim/src/heap.rs:229]:
```suggestion
        assert_eq!(heap.pages.len(), aligned.len() - 2);
```

File: enarx-keep-sgx-shim/src/heap.rs [hunk 4]
```diff
@@ -200,3 +204,156 @@ impl Heap {
...
+#[cfg(test)]
+mod tests {
+    use super::*;
+    const ERRNO_BASE: usize = !0xfff;
+
+    unsafe fn new_checked(bytes: &mut [u8]) -> Heap {
+        let aligned = slice(bytes.as_mut_ptr(), bytes.len())
+            .align_to_mut::<Page>()
+            .1;
+        let pages = aligned.len();
```
Comment (npmccallum) [inline enarx-keep-sgx-shim/src/heap.rs:217]:
```suggestion
```

File: enarx-keep-sgx-shim/src/heap.rs [hunk 5]
```diff
@@ -200,3 +204,128 @@ impl Heap {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    const ERRNO_BASE: usize = !0xfff;
+    const PROT: usize = libc::PROT_READ as _;
+    const FLAGS: usize = (libc::MAP_PRIVATE | libc::MAP_ANONYMOUS) as _;
+    const NUM_PAGES: usize = 10;
+
+    unsafe fn new_checked(bytes: &mut [u8]) -> Heap {
```
Comment (npmccallum) [inline enarx-keep-sgx-shim/src/heap.rs:216]: Why is this whole function unsafe?

**Comment by npmccallum** [comment]
@kubkon I agree completely on fuzzing. Care to open an issue for that?

**Comment by kubkon** [comment]
> @kubkon I agree completely on fuzzing. Care to open an issue for that?

Done: #718.

**Pull request merged**